### DISCUSSION TITLE
IPOPT performance can be much worse in 3.6.0

### DISCUSSION THREAD

**Discussion by thomasfork** [casadi/casadi#3076] [Q&A]
I am updating Python code that uses CasADi to version 3.6.0 and am seeing far worse (5x slower or failure to converge) performance of IPOPT on several problems that work fine in 3.5.5. They solve in ~10 seconds on 3.5.5 on Ubuntu 20.04 or Windows.

Time spent in NLP evaluations is <1 second on both versions.
Providing an HSL linear solver (MA97) is an improvement but does not resolve the slowdown. (ie. mumps on 3.5.5 is 5x faster)
I was able to reproduce this in version 3.5.6rc on Ubuntu 22.04.

I am looking for suggestions to debug this. (besides changing the optimization problem)
I can provide the nlp if useful. It is large so suggestions on exporting are welcome (serialized by casadi it was 240MB)

**Comment by jgillis** [comment]
Dear @thomasfork

I started a FAQ for your question here: https://github.com/casadi/casadi/wiki/FAQ:-How-to-debug-regressions-in-numerical-solvers-when-updating-CasADi%3F
Let me know if this helps.

Best regards,
  Joris

**Reply by thomasfork** [reply]
Thank you Joris, great to know these outputs exist.

After some more debugging the culprit was the newer version of MUMPS.
Seemingly this is a [known issue](https://coin-or.github.io/Ipopt/md_ChangeLog.html#:~:text=Added%20support%20for%20Mumps%205.2.x%2C%20though%20initial%20experiments%20on%20CUTEst%20indicated%20that%2C%20on%20average%2C%20performance%20is%20worse%20than%20when%20using%20Mumps%204.10.0.). The case where 3.6.0 was slow with HSL was isolated to that machine, it worked fine on several others.

Best Regards,
Thomas

**Comment by jgillis** [comment]
This cannot be the culprit, since Ipopt is configured with 4.0.3 cfr https://github.com/coin-or-tools/ThirdParty-Metis.git release/2.0.0

**Reply by thomasfork** [reply]
I'm not able to see the significance of your link, but when I run IPOPT with default options on casadi 3.6.0 I get the following output:
`This is Ipopt version 3.14.11, running with linear solver MUMPS 5.4.1.`

In casadi 3.5.5 it is ipopt version 3.12.3 which must have used an earlier version of MUMPS. (based on when it was released)

If I go further and set 'ipopt.print_level' : 7 and run the nlp with the stock MUMPS solver I get the following outputs:

In casadi v. 3.5.5:
```
**************************************************
*** Solving the Primal Dual System for Iteration 0:
**************************************************

Solving system with delta_x=0.000000e+00 delta_s=0.000000e+00
                    delta_c=0.000000e+00 delta_d=0.000000e+00
Calling MUMPS-2 for numerical factorization at cpu time     14.640 (wall      0.157).
Done with MUMPS-2 for numerical factorization at cpu time     14.684 (wall      0.201).
Number of doubles for MUMPS to hold factorization (INFO(9)) = 1342823
Number of integers for MUMPS to hold factorization (INFO(10)) = 194322
In MumpsSolverInterface::Factorization: negevals_ = 7928, but numberOfNegEVals = 7903
```
and in casadi v. 3.6.0
```
**************************************************
*** Solving the Primal Dual System for Iteration 0:
**************************************************

Solving system with delta_x=0.000000e+00 delta_s=0.000000e+00
                    delta_c=0.000000e+00 delta_d=0.000000e+00
Calling MUMPS-2 for numerical factorization.
Done with MUMPS-2 for numerical factorization.
Number of doubles for MUMPS to hold factorization (INFO(9)) = 2851031
Number of integers for MUMPS to hold factorization (INFO(10)) = 183162
In MumpsSolverInterface::Factorization: negevals_ = 7928, but numberOfNegEVals = 7903
```

There is an enormous increase in the number of doubles MUMPS is using for the factorization, and over many (much slower) iterations it doesn't converge. Both points seem to hold only on certain nlp's, the one here is rather difficult. 

If I switch to HSL the shown steps (and following steps) are identical across the two casadi versions, down the the total number of iterations and time needed to converge (144 iterations and 38 seconds on the small laptop I just tested this on)

**Comment by jgillis** [comment]
Oh my bad, I misread. I thought we were talking about METIS, a graph sorting routine within MUMPS.
